common/rpc/direct_peer_chooser.go

```diff
@@ -22,93 +22,244 @@ package rpc
-	c, ok := g.getLegacyChooser()
-	if ok {
-		return c.Stop()
+	if !atomic.CompareAndSwapInt32(&g.status, common.DaemonStatusStarted, common.DaemonStatusStopped) {
+		return nil
+	}
+
+	var err error
+	// Stop legacy chooser if it was initialized
+	if g.legacyChooser != nil {
+		err = g.legacyChooser.Stop()
 	}
 
-	return nil // no-op
+	// Release all peers if there's any
+	g.updatePeersInternal(nil)
+
+	g.logger.Info("direct peer chooser stopped", tag.Error(err))
+	return err
 }
 
 // IsRunning statisfies the peer.Chooser interface.
 func (g *directPeerChooser) IsRunning() bool {
-	c, ok := g.getLegacyChooser()
-	if ok {
-		return c.IsRunning()
+	if atomic.LoadInt32(&g.status) != common.DaemonStatusStarted {
+		return false
+	}
+
+	if !g.enableConnRetainMode() {
+		c, ok := g.getLegacyChooser()
+		if ok {
+			return c.IsRunning()
+		}
+		return false
 	}
 
 	return true // no-op
 }
 
 // Choose returns an existing peer for the shard key.
+// ShardKey is {host}:{port} of the peer. It could be tchannel or grpc address.
 func (g *directPeerChooser) Choose(ctx context.Context, req *transport.Request) (peer peer.Peer, onFinish func(error), err error) {
-	if g.enableConnRetainMode != nil && !g.enableConnRetainMode() {
+	if !g.enableConnRetainMode() {
 		return g.chooseFromLegacyDirectPeerChooser(ctx, req)
 	}
 
 	if req.ShardKey == "" {
 		return nil, nil, yarpcerrors.InvalidArgumentErrorf("chooser requires ShardKey to be non-empty")
 	}
 
-	// TODO: implement connection retain mode
-	return nil, nil, yarpcerrors.UnimplementedErrorf("direct peer chooser conn retain mode unimplemented")
+	g.mu.RLock()
+	p, ok := g.peers[req.ShardKey]
+	if ok {
+		g.mu.RUnlock()
+		return p, func(error) {}, nil
+	}
+	g.mu.RUnlock()
+
+	// peer is not cached, add new peer
+	p, err = g.addPeer(req.ShardKey)
+	if err != nil {
+		return nil, nil, yarpcerrors.InternalErrorf("failed to add peer for shard key %v, err: %v", req.ShardKey, err)
+	}
+
+	return p, func(error) {}, nil
 }
 
-func (g *directPeerChooser) UpdatePeers(members []membership.HostInfo) {
-	// TODO: implement
+// UpdatePeers removes peers that are not in the members list.
+// Do not create actual yarpc peers for the members. They are created lazily when a request comes in (Choose is called).
+func (g *directPeerChooser) UpdatePeers(serviceName string, members []membership.HostInfo) {
+	if g.serviceName != serviceName {
+		g.logger.Debug("This is not the service chooser is created for. Ignore such updates.", tag.Dynamic("members-service", serviceName))
+		return
+	}
+
 	g.logger.Debug("direct peer chooser got a membership update", tag.Counter(len(members)))
+
+	// If the chooser is not started, do not act on membership changes.
+	// If membership updates arrive after chooser is stopped, ignore them.
+	if atomic.LoadInt32(&g.status) != common.DaemonStatusStarted {
+		return
+	}
+
+	g.updatePeersInternal(members)
+}
+
+func (g *directPeerChooser) updatePeersInternal(members []membership.HostInfo) {
```

Comment:
this looks to not actually update (it's lazy, like you said), it's instead trimming no longer valid peers? Would naming it something something indicating this make sense?

Reply:
Zijian brought same point. I will address in a follow up since this is merged and prerelease is created